storage/transfermanager/integration_test.go

```diff
@@ -0,0 +1,851 @@
+}
+
+// TODO: once we provide a DownloaderBuffer that implements WriterAt in the
+// library, we can use that instead.
+type testWriter struct {
+	b      []byte
+	crc32c uint32
+	bufs   [][]byte // temp bufs that will be joined on Close()
+}
+
+// Close must be called to finalize the buffer
+func (tw *testWriter) Close() error {
+	tw.b = bytes.Join(tw.bufs, nil)
+	crc := crc32.New(crc32.MakeTable(crc32.Castagnoli))
+
+	_, err := io.Copy(crc, bytes.NewReader(tw.b))
+	tw.crc32c = crc.Sum32()
+	return err
+}
+
+func (tw *testWriter) WriteAt(b []byte, offset int64) (n int, err error) {
+	// TODO: use the offset. This is fine for now since reads are not yet sharded.
+	copiedB := make([]byte, len(b))
+	copy(copiedB, b)
+	tw.bufs = append(tw.bufs, copiedB)
+
+	return len(b), nil
+}
+
+func deleteExpiredBuckets(prefix string) error {
+	if testing.Short() {
+		return nil
+	}
+
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return fmt.Errorf("NewClient: %v", err)
+	}
+
+	projectID := testutil.ProjID()
+	it := client.Buckets(ctx, projectID)
+	it.Prefix = prefix
+	for {
+		bktAttrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if time.Since(bktAttrs.Created) > bucketExpiryAge {
+			log.Printf("deleting bucket %q, which is more than %s old", bktAttrs.Name, bucketExpiryAge)
+			if err := killBucket(ctx, client, bktAttrs.Name); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// killBucket deletes a bucket and all its objects.
+func killBucket(ctx context.Context, client *storage.Client, bucketName string) error {
+	bkt := client.Bucket(bucketName)
+	// Bucket must be empty to delete.
+	it := bkt.Objects(ctx, nil)
+	for {
+		objAttrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if err := bkt.Object(objAttrs.Name).Delete(ctx); err != nil {
+			return fmt.Errorf("deleting %q: %v", bucketName+"/"+objAttrs.Name, err)
+		}
+	}
+	// GCS is eventually consistent, so this delete may fail because the
+	// replica still sees an object in the bucket. We log the error and expect
+	// a later test run to delete the bucket.
+	if err := bkt.Delete(ctx); err != nil {
+		log.Printf("deleting %q: %v", bucketName, err)
+	}
+	return nil
+}
+
+// downloadTestBucket provides a bucket that can be reused for tests that only
+// download from the bucket.
+type downloadTestBucket struct {
+	bucket        string
+	objects       []string
+	contentHashes map[string]uint32
+	objectSize    int64
+}
+
+// Create initializes the downloadTestBucket, creating a bucket and populating
+// objects in it. All objects are of the same size but with different contents
+// and can be mapped to their respective crc32c hash in contentHashes.
+func (tb *downloadTestBucket) Create(prefix string) error {
+	if testing.Short() {
+		return nil
+	}
+	ctx := context.Background()
+
+	tb.bucket = prefix + uidSpace.New()
```

Comment:
I think we should talk about adding some more objects (at least more than the thread count) and of a wider variety of sizes. Can be in a follow up PR though.